Some enhancements

sagotch opened the pull request:
Summary of these improvements:
- `function` and `eval`: avoid useless buffering and marshaling of values andreturn the last value only. Previous implementation returned the first value only, but this one is more consistent with OCaml, and functions/eval must return only one value anyway.
- Added some filters: `nth` `forall`
- Support array/list appending with the `+` operator. Consistent with the fatch that it already is a multi-purpose operateur used to concat strings.
- Some fixes in the lexer (whitespace control and empty text at the end of parsing) and better error messages.
- Regex: switched from `Re.Pcre` to `Re.Str` so we can use groups in the replacement string.
- Fixed `jg_cmdlin` and include `jingoo` binaray in installation.
- Rework of `attr` filter
- Fixed `jg_slice` (which is not the same as `jg_batch`)
- Renamed `jg_filter` to `jg_select` because "filter" is used too often to speak of a function or of the filter statement.

The rest is mainly cosmetic.

tategakibunko commented:
Hmm, I've looked jinja2 impl of `slice` just now, and you are correct, thank you!

https://gemfury.com/offerpop/python:Jinja2/-/content/jinja2/filters.py

New runtime function(`jg_fold`) is jingoo original filter, and it's unicode safe, Great!

But unit test of `jg_fold`(for Tstr) is not implemented yet, so I'll try to add one later.

And as you pointed out, I also want `func_arg_xxx`(xxx for general number), but I don't know how to implement it. If you have any good idea, feel free to update.

By the way, I can't find any document about `OUNIT_TEST_DATA_DIR`, where did you find it?

tategakibunko commented:
I added unit test for jg_fold using Tstr, simple ascii is good, but error if it's used against multi-byte strings.

```bash
```

sagotch commented:
> By the way, I can't find any document about OUNIT_TEST_DATA_DIR, where did you find it?

I guess that I found it here but I have to say I do not remember... http://ounit.forge.ocamlcore.org/api-ounit/OUnit2.Conf.html

> I added unit test for jg_fold using Tstr, simple ascii is good, but fails if it's used against multi-byte strings.

You're right. A stupid mistake where I partially forgot about how UTF8.sub actually do what should be done with utf8 stuff, Thank you for spotting this!
Fixed in https://github.com/tategakibunko/jingoo/pull/68

sagotch commented:
> And as you pointed out, I also want func_arg_xxx(xxx for general number), but I don't know how to implement it. If you have any good idea, feel free to update.

I have something, but it is not working with varargs functions. This problem only be a problem for `loop.cycle`. If it is okay to say that it should take a single sequence element as paramater it is okay.

`loop.cycle ("even", "odd")` would become `loop.cycle (["even", "odd"])`.

Note that this change is just as well, as it also allow to use a value as argument of the `loop.cycle` function `loop.cycle (x)` with could be useful in macro, or if you have multiple occurrence of a `loop.cycle` with the same values. As a side effect, it also allow you to use `loop.cycle ("abcd")` as a shorthand for `loop.cycle (["a","b", "c","d"])` (or `loop.cycle ("a","b", "c","d")`) with the previous implementations.

I need to clean the code, but I'll submit a PR soon so you can tell me about it.
